radixdlt-core/radixdlt/src/main/java/com/radixdlt/api/construction/LogCMStore.java

```diff
@@ -61,40 +66,39 @@ public boolean isVirtualDown(Transaction dbTxn, SubstateId substateId) {
 	}
 
 	@Override
-	public Optional<Particle> loadAddr(Transaction tx, REAddr rri) {
-		return readableAddrs.loadAddr(tx, rri);
+	public Optional<Particle> loadAddr(Transaction tx, REAddr rri, SubstateDeserialization deserialization) {
+		return readableAddrs.loadAddr(tx, rri, deserialization);
 	}
 
 	@Override
-	public Optional<Particle> loadUpParticle(Transaction dbTxn, SubstateId substateId) {
+	public Optional<Particle> loadUpParticle(Transaction dbTxn, SubstateId substateId, SubstateDeserialization deserialization) {
 		var txnId = substateId.getTxnId();
 		return txnIndex.get(txnId)
 			.flatMap(txn -> {
-				var buf = ByteBuffer.wrap(txn.getPayload());
 				var index = substateId.getIndex().orElseThrow();
-				int cur = 0;
-				while (buf.hasRemaining()) {
-					try {
-						var i = REInstruction.readFrom(txn, cur, buf);
-						if (cur == index) {
-							if (i.getMicroOp() != REInstruction.REMicroOp.UP) {
-								return Optional.empty();
-							}
-							Substate s = i.getData();
-							return Optional.of(s.getParticle());
-						}
-					} catch (DeserializeException e) {
-						throw new IllegalStateException("Cannot deserialize instruction@" + cur, e);
+				try {
+					var instructions = reParser.parse(txn).instructions();
```

Comment:
As an optimisation opportunity for future, we might consider building index for up particles so no full transaction parsing is required.